File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;


public class Main {
    static int[] counts;
    static int n;
    static List<Integer>[] nodes;
    static boolean[] visit;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        counts = new int[n + 1];
        nodes = new List[n + 1];
        visit = new boolean[n + 1];
        for (int i = 0; i < n + 1; i++) {
            nodes[i] = new ArrayList<>();
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            nodes[b].add(a);
        }

        for (int i = 1; i < n + 1; i++) {
            counts[i] = bfs(i) ;

        }
//        System.out.println(Arrays.toString(counts));
        int maxV = Arrays.stream(counts).max().getAsInt();
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < n+1; i++) {
            if (counts[i] == maxV) {
                sb.append(i).append(" ");
            }
        }
        System.out.println(sb);
    }

    private static int bfs(int current) {
        Queue<Integer> q = new ArrayDeque<>();
        q.add(current);
        boolean[] visit = new boolean[n + 1];
        int count = 0;
        while (!q.isEmpty()) {
            int tmp = q.poll();
            if (visit[tmp]){
                continue;
            }
            visit[tmp] = true;
            count++;
            q.addAll(nodes[tmp]);
        }
        return count;
    }


}

File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.cc
#include <iostream>
#include <vector>
#include <queue>
#include <sstream>

using namespace std;

const int MOD = 1e9 + 9;
const int INF = 2e9;

int n, m, k, x, y;
vector<vector<int>> arr;
vector<int> visit;
vector<int> counts;
vector<vector<int>> nodes;

vector<int> dx = {0, 0, 1, -1};
vector<int> dy = {1, -1, 0, 0};
stringstream sb;

int bfs(int current) {
    queue<int> q;
    q.push(current);
    vector<bool> visited(n + 1, false);
    visited[current] = true;
    int count = 0;
    
    while (!q.empty()) {
        int tmp = q.front();
        q.pop();
        
        for (int next : nodes[tmp]) {
            if (visited[next]) {
                continue;
            }
            visited[next] = true;
            count++;
            q.push(next);
        }
    }
    
    return count;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> m;
    counts.resize(n + 1);
    nodes.resize(n + 1);

    for (int i = 0; i < n + 1; i++) {
        nodes[i] = vector<int>();
    }

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        nodes[b].push_back(a);
    }

    int maxCount = 0;
    for (int i = 1; i < n + 1; i++) {
        counts[i] = bfs(i);
        maxCount = max(counts[i], maxCount);
    }

    for (int i = 1; i < n + 1; i++) {
        if (counts[i] == maxCount) {
            sb << i << " ";
        }
    }

    cout << sb.str() << endl;

    return 0;
}